doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,360 @@
+    collections can grow without bounds. We therefore don't want to enable
+    sinking to persist for long-running Materialize deployments (like cloud). To
+    still be able to collect introspection data for such deployments, we need to
+    keep the existing mechanism of peeking introspection arrangements around
+    (even if it only works reliably for single-replica clusters).
+  * Once we have sinks working for replicated clusters, we'll want to think
+    about using them to sink the introspection arrangements instead of the
+    hand-rolled solution proposed in this design.
+
+To sink introspection data, replicas need the following storage connection
+information:
+
+  * Blob URI
+  * Consensus URI
+  * A shard ID for every log variant
+
+This connection information is passed to replicas through the
+`CreateInstance` compute command.
+
+#### Changes to the Compute API
+
+The `LoggingConfig` contained in `ComputeCommand::CreateInstance` is extended to
+also contain the storage connection information mentioned above (details subject
+to bike-shedding):
+
+```rust
+pub struct LoggingConfig {
+    pub granularity_ns: u128,
+    pub active_logs: HashMap<LogVariant, ActiveLog>,
+    pub log_logging: bool,
+    pub blob_uri: String,
+    pub consensus_uri: String,
+}
+
+pub struct ActiveLog {
+    pub index_id: GlobalId,
+    pub shard_id: ShardId,
+}
+```
+
+The compute controller generally assumes that it can send the same commands to
+all replicas in a cluster. This will not be the case anymore with the updated
+`LoggingConfig`, since shard IDs are specific to replicas. Therefore, the
+compute controller must take care of adjusting `CreateInstance` commands to the
+respective target replicas before sending them.
+
+#### Cleanup on Replica Restart
+
+When a replica is restarted for whatever reason, we need to make sure that the
+persist shards they log their introspection data to are cleared. Otherwise,
+we might end up with stale introspection data that was written by a previous
+incarnation of the replica and never retracted because subsequent incarnations
+were not aware of the stale data's existence.
+
+Replicas should clear their introspection shards as soon as they receive a
+`CreateInstance` command, before they initialize the introspection dataflows.
+
+`persist` does not offer a convenient API to clear shards, but it should
+still provide all necessary means. To clear a shard, we can:
+
+  1. Make a snapshot at the current time.
+  1. Issue retraction updates for each record in the snapshot.
+
+Note that this way of clearing shards in-place will not work if an old
+incarnation of the replica is still around logging new data to the same shard,
+as the presence of multiple writers can lead to inconsistent/undefined shard
+contents. This issue will be resolved in the future by the use of storage
+renditions.
+
+#### Cleanup on Replica Drop
+
+When a replica receives a `DropInstance` command, it should drop all its
```

Comment:
Yes, this is mainly a placeholder because I wasn't clear on how cleaning up a collection looks like. Will update this now that Dan has answered that question.

Reply:
I've removed this section for now. Someone needs to signal the storage collections' end-of-life by downgrading the `since`s to the empty frontier. That should be the same component also responsible for making sure `since` gets continually advanced during the collections' lifetimes. But I'm not sure yet what component this should be (maybe the replica, maybe the storage controller).